### ISSUE TITLE
UnsupportedOperationException: dataType

### ISSUE THREAD

**Issue by arijit999** [AbsaOSS/spline#624]
```
20/04/02 20:58:29 INFO FileFormatWriter: Finished processing stats for write job 55b3e202-a211-4af3-a5ad-4b8dbbe0a945.
20/04/02 20:58:29 INFO SparkLineageInitializer$: Spline v0.4.2 is initializing...
20/04/02 20:58:29 INFO SparkLineageInitializer$: Spline successfully initialized. Spark Lineage tracking is ENABLED.
20/04/02 20:58:30 WARN ExecutionListenerManager: Error executing query execution listener
java.lang.UnsupportedOperationException: dataType
	at org.apache.spark.sql.catalyst.expressions.WindowSpecDefinition.dataType(windowExpressions.scala:54)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.getDataType(ExpressionConverter.scala:77)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:71)
	at za.co.absa.spline.harvester.converter.ExpressionConverter$$anonfun$convert$4.apply(ExpressionConverter.scala:72)
	at za.co.absa.spline.harvester.converter.ExpressionConverter$$anonfun$convert$4.apply(ExpressionConverter.scala:72)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:296)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:72)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:41)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$1$$anonfun$apply$1.applyOrElse(OperationParamsConverter.scala:42)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$1$$anonfun$apply$1.applyOrElse(OperationParamsConverter.scala:35)
	at scala.PartialFunction$OrElse.apply(PartialFunction.scala:167)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$za$co$absa$spline$harvester$converter$ValueDecomposer$$recursion$1.apply(ValueDecomposer.scala:39)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$za$co$absa$spline$harvester$converter$ValueDecomposer$$recursion$1.apply(ValueDecomposer.scala:39)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$2$$anonfun$apply$1$$anonfun$applyOrElse$4.apply(ValueDecomposer.scala:51)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$2$$anonfun$apply$1.applyOrElse(ValueDecomposer.scala:51)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$2$$anonfun$apply$1.applyOrElse(ValueDecomposer.scala:44)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$1$$anonfun$apply$1.applyOrElse(OperationParamsConverter.scala:35)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$1$$anonfun$apply$1.applyOrElse(OperationParamsConverter.scala:35)
	at scala.PartialFunction$OrElse.apply(PartialFunction.scala:167)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$za$co$absa$spline$harvester$converter$ValueDecomposer$$recursion$1.apply(ValueDecomposer.scala:39)
	at za.co.absa.spline.harvester.converter.ValueDecomposer$$anonfun$za$co$absa$spline$harvester$converter$ValueDecomposer$$recursion$1.apply(ValueDecomposer.scala:39)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$convert$4.apply(OperationParamsConverter.scala:58)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$convert$4.apply(OperationParamsConverter.scala:54)
	at scala.collection.TraversableLike$WithFilter$$anonfun$map$2.apply(TraversableLike.scala:683)
	at scala.collection.immutable.Map$Map4.foreach(Map.scala:188)
	at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:682)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter.convert(OperationParamsConverter.scala:54)
	at za.co.absa.spline.harvester.builder.GenericNodeBuilder.build(GenericNodeBuilder.scala:34)
	at za.co.absa.spline.harvester.builder.GenericNodeBuilder.build(GenericNodeBuilder.scala:24)
	at za.co.absa.spline.harvester.LineageHarvester$$anonfun$harvest$1$$anonfun$4.apply(LineageHarvester.scala:76)
	at za.co.absa.spline.harvester.LineageHarvester$$anonfun$harvest$1$$anonfun$4.apply(LineageHarvester.scala:76)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:296)
	at za.co.absa.spline.harvester.LineageHarvester$$anonfun$harvest$1.apply(LineageHarvester.scala:76)
	at za.co.absa.spline.harvester.LineageHarvester$$anonfun$harvest$1.apply(LineageHarvester.scala:69)
	at scala.Option.flatMap(Option.scala:171)
	at za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:69)
	at za.co.absa.spline.harvester.QueryExecutionEventHandler.onSuccess(QueryExecutionEventHandler.scala:41)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1.apply(SplineQueryExecutionListener.scala:37)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener$$anonfun$onSuccess$1.apply(SplineQueryExecutionListener.scala:37)
	at scala.Option.foreach(Option.scala:257)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:37)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(QueryExecutionListener.scala:124)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1$$anonfun$apply$mcV$sp$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling$1.apply(QueryExecutionListener.scala:145)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling$1.apply(QueryExecutionListener.scala:143)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at org.apache.spark.sql.util.ExecutionListenerManager.org$apache$spark$sql$util$ExecutionListenerManager$$withErrorHandling(QueryExecutionListener.scala:143)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply$mcV$sp(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager$$anonfun$onSuccess$1.apply(QueryExecutionListener.scala:123)
	at org.apache.spark.sql.util.ExecutionListenerManager.readLock(QueryExecutionListener.scala:156)
	at org.apache.spark.sql.util.ExecutionListenerManager.onSuccess(QueryExecutionListener.scala:122)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:678)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:285)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:271)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:229)
	at org.apache.spark.sql.DataFrameWriter.parquet(DataFrameWriter.scala:566)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
20/04/02 20:58:30 INFO SparkContext: Invoking stop() from shutdown hook
```

**Comment by wajda** [comment]
Could you please provide a Spark app snippet that reproduces the issue?

**Comment by wajda** [comment]
What Spark version are you using?

**Comment by cerveada** [comment]
- [WindowSpecDefinition.dataType](https://github.com/apache/spark/blob/branch-2.4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala) always throws.

- This expression is most probably created when using `column.over(...)` function.
- Since `WindowSpecDefinition` extends `Unevaluable` that doesn't suppose to live past analysis or optimization, the solution may be to ignore it.

**Comment by cerveada** [comment]
We will release fix for 0.5.x that will allow agent to gather the data properly. 

Attribute lineage feature will not take into account the information from windowExpressions - this will be fixed in spline 0.6 as #668
